clients/rust/examples/sysadmin-example.rs

```diff
@@ -0,0 +1,81 @@
+extern crate sysadmin_bindings;
+extern crate protobuf;
+
+use std::time::Duration;
+use std::net::{IpAddr, Ipv4Addr, SocketAddr};
+use sysadmin_bindings::sysadminctl::{ConfigValue, Set, Get, Commit, CommitConfig, Response};
+
+
+/// Available Payloads:
+///     Set
+///     Get
+///     Commit
+///     Drop
+///     FireHooks
+///     EraseKey
+///     Rollback
+///     Reset
+///     DumpHooks
+///     TriggerHook
+///     Blame
+///     InFlight
+
+fn make_client() -> sysadmin_bindings::SysAdminClient {
+    let default_timeout = Duration::new(2, 0);
+    let socket = SocketAddr::new(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)), 4000_u16);
+    let sysadminclient =
+        sysadmin_bindings::SysAdminClient::new(socket, default_timeout, 0_u32, 1_u32);
+    sysadminclient
+}
+
+fn make_set() -> Set {
+    let mut set = Set::new();
+    set.set_key("fizz".to_string());
+    let mut cv = ConfigValue::new();
+    cv.set_int32val(123_i32);
+    set.set_value(cv);
+    set
+}
+
+fn make_commit() -> Commit {
+    let mut commit = Commit::new();
+    commit.set_config(CommitConfig::DEFAULT);
+    commit
+}
+
+fn make_get() -> Get {
+    let mut get = Get::new();
+    get.set_key("fizz".to_string());
+    get
+}
+
+fn print_resp(resp: Result<Response, protobuf::error::ProtobufError>) {
+    match resp {
+        Ok(r) => println!("Response = {:?}", r),
+        Err(e) => println!("Err = {:?}", e),
+    }
+}
+
+fn main() {
+    // Build the sysadmin server and run it with:
+    //      ./build/src/sysadmin test/configs/config.yaml
+
+
+    let mut sysadminclient = make_client();
+    let set = make_set();
```

Comment:
Could we make these methods on the actual client? Building up messages manually feels painful when the API is so simple. Ideally I'd think you would be able to just do something like this:
```
let client = SysAdminClient::new();
client.set("key", 123);
client.commit();
let response = client.get("key");
let value = response.value::<int>().unwrap();
```

Reply:
I spent some time looking for a way to add that to the client, but ultimately found that you have to specify the value with a type-specific function for `ConfigValue`. E.g. 
```
    let mut cv = ConfigValue::new();
    cv.set_int32val(123_i32);
    set.set_value(cv);
```
The rust-protobuf lib wants you to be pretty explicit, and it restricts what you can do in the name of "object safety".  It can be a bit clunky, but like you said, it's simple. 

For now I don't think that I can, or should, attempt further abstractions over the generated protobuf module. I'm not that experienced with Sysadmin, but as time goes on I might be able to identify some sensible helpers which can be added to the lib. E.g. Commit, Inflight, Get (since it always takes a String)

Reply:
The problem I have with the current API design is the library API is more complicated than the wire API. The point behind providing sysadmin libraries is to make user space code as simple as possible. A consumer of this library should have no idea that it is using protobufs, but the `ConfigValue` struct is exposed in the public API - this makes changing the underlying protobufs impossible without breaking user code. The protobuf API is explicit because it is generated code - it has no understanding of the application level use of those messages so they can't generate a smarter interface to the messages. In order to make this API usable we need to abstract that away and create the smarter interface that we would want to use in our applications.

It feels like what is ultimately needed is a layer on top of `ConfigValue`. There are a few ways you could do this. One way is that you could have an enum `SysadminValue` that represents all of the different types of sysadmin values. You can implement `From<T>` for `SysadminValue` for whatever rust types we want to support. Then your `set` method could look something like this:
```
fn set(&self, key: &str, value: V) -> Result 
    where V: Into<SysadminValue>
```
You could then do something similar for unpacking the get response.

You could also create your own `IntoSysadminValue` trait that works similarly.

Reply:
> It feels like what is ultimately needed is a layer on top of ConfigValue. There are a few ways you could do this.

I took a look back at the original .proto file and can see what you mean. It should be possible and worthwhile to simplify the interface.